process/sync/baseSync.go

```diff
@@ -298,25 +294,46 @@ func (boot *baseBootstrap) ShouldSync() bool {
 	}
 	boot.statusHandler.SetUInt64Value(core.MetricIsSyncing, result)
 
+	go boot.requestHeadersIfSyncIsStuck()
+
 	return !isNodeSynchronized
 }
 
-func (boot *baseBootstrap) trySyncTrigger() {
+func (boot *baseBootstrap) requestHeadersIfSyncIsStuck() {
+	if !boot.isNodeSynchronized {
+		return
+	}
+
 	currHeader := boot.chainHandler.GetCurrentBlockHeader()
 	if currHeader == nil {
 		return
 	}
 
 	roundDiff := uint64(boot.rounder.Index()) - currHeader.GetRound()
-	if roundDiff < tryRequestHeaderDelta {
+	if roundDiff <= process.MaxRoundsWithoutNewBlockReceived {
 		return
 	}
 
-	shouldTrySync := roundDiff%tryRequestHeaderDelta == 0
+	numRequestedHdrs := 0
+	nonce := boot.getNonceForNextBlock()
```

Comment:
is this concurrent safe?

Reply:
If this is not concurrent safe that means we have a problem on each call from chainHandler. This is the method listed:

// getNonceForNextBlock will get the nonce for the next block we should request
func (boot *baseBootstrap) getNonceForNextBlock() uint64 {
	nonce := uint64(1) // first block nonce after genesis block
	if boot.chainHandler != nil && boot.chainHandler.GetCurrentBlockHeader() != nil {
		nonce = boot.chainHandler.GetCurrentBlockHeader().GetNonce() + 1
	}
	return nonce
}

Reply:
check.ifnil

Reply:
Refactored